### PR TITLE
Linearly rescaled constraint parameters

### PR DESCRIPTION
**Pull request by VMatthijs**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary
This is the Language part of a solution to the issue:
stan-dev/stan#2656 .

#### Intended Effect
Make non-centered parameterisation more convenient.

#### How to Verify
Run included tests and test-models.

#### Documentation

See reference manual update.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by VMatthijs** [comment]
This is ready for review once the corresponding math PR (https://github.com/stan-dev/math/pull/1048) has been merged.

@mitzimorris , would you maybe be able to do this?

**Comment by mitzimorris** [comment]
yes!

File: src/stan/lang/grammars/common_adaptors_def.hpp [hunk 1]
```diff
@@ -9,5 +9,9 @@ BOOST_FUSION_ADAPT_STRUCT(stan::lang::range,
                           (stan::lang::expression, low_)
                           (stan::lang::expression, high_) )
 
+BOOST_FUSION_ADAPT_STRUCT(stan::lang::locscale,
```
[the remark this answers is not in the document]
Comment (VMatthijs) [reply src/stan/lang/grammars/common_adaptors_def.hpp]: Done.

File: src/stan/lang/grammars/semantic_actions_def.cpp [hunk 2]
```diff
@@ -3215,7 +3244,7 @@ namespace stan {
 
     // only used to debug grammars
     void trace::operator()(const std::string& msg) const {
-      //      std::cout << msg << std::endl;
+            std::cout << msg << std::endl;
```
[the remark this answers is not in the document]
Comment (VMatthijs) [reply src/stan/lang/grammars/semantic_actions_def.cpp]: Done.

File: src/stan/io/reader.hpp [hunk 3]
```diff
@@ -676,6 +676,57 @@ namespace stan {
...
+       */
+      template <typename TL, typename TS>
+      inline T scalar_locscale(const TL loc, const TS scale) {
+        T x(scalar());
+        return x;
+      }
+
+      /**
+       * Return the next scalar transformed to be between
+       * the specified lower and scales.
```
[the remark this answers is not in the document]
Comment (VMatthijs) [reply src/stan/io/reader.hpp]: Done.

File: src/test/test-models/bad/validate_set_double_locscale_scale_bad.stan [hunk 4]
```diff
@@ -0,0 +1,8 @@
+parameters {
```
[the remark this answers is not in the document]
Comment (VMatthijs) [reply src/test/test-models/bad/validate_set_double_locscale_scale_bad.stan]: Done.

**Comment by VMatthijs** [comment]
I've made the changes you requested, @mitzimorris . :-)

**Comment by mitzimorris** [comment]
needs cpplint fix

**Comment by VMatthijs** [comment]
I think this can be merged now once it passes tests.

**Comment by mitzimorris** [comment]
tests still failing - from error log:
```
In file included from src/stan/lang/ast_def.cpp:23:
src/stan/lang/ast/type/double_block_type_def.hpp:14:11: error: invalid operands to binary expression ('const char *' and 'const char *')
          + "a location/scale.");
          ^ ~~~~~~~~~~~~~~~~~~~
In file included from src/stan/lang/ast_def.cpp:28:
src/stan/lang/ast/type/matrix_block_type_def.hpp:16:11: error: invalid operands to binary expression ('const char *' and 'const char *')
          + "a location/scale.");
          ^ ~~~~~~~~~~~~~~~~~~~
In file included from src/stan/lang/ast_def.cpp:33:
src/stan/lang/ast/type/row_vector_block_type_def.hpp:15:11: error: invalid operands to binary expression ('const char *' and 'const char *')
          + "a location/scale.");
          ^ ~~~~~~~~~~~~~~~~~~~
In file included from src/stan/lang/ast_def.cpp:38:
src/stan/lang/ast/type/vector_block_type_def.hpp:14:11: error: invalid operands to binary expression ('const char *' and 'const char *')
          + "a location/scale.");
          ^ ~~~~~~~~~~~~~~~~~~~
4 errors generated.
```

**Comment by VMatthijs** [comment]
Sorry, @mitzimorris ! I clearly broke something while trying to fix the cpplint. It should be working now.

**Comment by VMatthijs** [comment]
Sorry, @mitzimorris ! I clearly broke something while trying to fix the cpplint. It should be working now.

**Comment by VMatthijs** [comment]
Sorry, @mitzimorris ! I clearly broke something while trying to fix the cpplint. It should be working now.

**Comment by VMatthijs** [comment]
@mitzimorris , could you maybe approve again?

**Comment by VMatthijs** [comment]
@mitzimorris , @seantalts , one of my Travis jobs keeps exceeding the maximum allowed time. Has that happened to you before? It doesn't seem to have anything to do with the current PR though...

**Comment by seantalts** [comment]
Yeah, Travis regularly does that :( Can you try restarting it? When you click on the job you should have permission to just restart the one that timed out I think...?

**Comment by VMatthijs** [comment]
Tried it a few times and it keeps timing out. Will try again.

**Comment by VMatthijs** [comment]
@mitzimorris , could you maybe merge this?

**Comment by VMatthijs** [comment]
Thanks!

**Pull request merged**